Sources/BrowserServicesKit/FeatureFlagger/ExperimentCohortsManager.swift

```diff
@@ -0,0 +1,146 @@
+    public let parentID: String
+    public let cohortID: String
+    public let enrollmentDate: Date
+}
+
+public typealias Experiments = [String: ExperimentData]
+
+public protocol ExperimentCohortsManaging {
+    /// Retrieves all the experiments a user is enrolled in
+    var experiments: Experiments? { get }
+
+    /// Resolves the cohort for a given experiment subfeature.
+    ///
+    /// This method determines whether the user is currently assigned to a valid cohort
+    /// for the specified experiment. If the assigned cohort is valid (i.e., it matches
+    /// one of the experiment's defined cohorts), the method returns the assigned cohort.
+    /// Otherwise, the invalid cohort is removed, and a new cohort is assigned if
+    /// `isAssignCohortEnabled` is `true`.
+    ///
+    /// - Parameters:
+    ///   - experiment: The `ExperimentSubfeature` representing the experiment and its associated cohorts.
+    ///   - isAssignCohortEnabled: A Boolean value indicating whether cohort assignment is allowed
+    ///     if the user is not already assigned to a valid cohort.
+    ///
+    /// - Returns: The valid `CohortID` assigned to the user for the experiment, or `nil`
+    ///   if no valid cohort exists and `isAssignCohortEnabled` is `false`.
+    ///
+    /// - Behavior:
+    ///   1. Retrieves the currently assigned cohort for the experiment using the `subfeatureID`.
+    ///   2. Validates if the assigned cohort exists within the experiment's cohort list:
+    ///      - If valid, the assigned cohort is returned.
+    ///      - If invalid, the cohort is removed from storage.
+    ///   3. If cohort assignment is enabled (`isAssignCohortEnabled` is `true`), a new cohort
+    ///      is assigned based on the experiment's cohort weights and saved in storage.
+    ///      - Cohort assignment is probabilistic, determined by the cohort weights.
+    ///
+    func resolveCohort(for experiment: ExperimentSubfeature, isAssignCohortEnabled: Bool) -> CohortID?
+}
+
+public class ExperimentCohortsManager: ExperimentCohortsManaging {
+
+    private var store: ExperimentsDataStoring
+    private let randomizer: (Range<Double>) -> Double
+    private let queue = DispatchQueue(label: "com.ExperimentCohortsManager.queue")
+
+    public var experiments: Experiments? {
+        get {
+            queue.sync {
+                store.experiments
+            }
+        }
+    }
+
+    public init(store: ExperimentsDataStoring = ExperimentsDataStore(), randomizer: @escaping (Range<Double>) -> Double = Double.random(in:)) {
+        self.store = store
+        self.randomizer = randomizer
+    }
+
+    public func resolveCohort(for experiment: ExperimentSubfeature, isAssignCohortEnabled: Bool) -> CohortID? {
+        queue.sync {
+            let assignedCohort = cohort(for: experiment.subfeatureID)
+            if experiment.cohorts.contains(where: { $0.name == assignedCohort }) {
+                return (assignedCohort)
+            } else {
```

Comment:
this else if not needed, because we are returning in if, so:
if experiment.cohorts.contains(where: { $0.name == assignedCohort }) {
    return (assignedCohort)
}
removeCohort(from: experiment.subfeatureID)
return isAssignCohortEnabled ? assignCohort(to: experiment) : nil